enhancements/hypershift/pluggable-konnectivity.md

```diff
@@ -0,0 +1,159 @@
+  - csrwng
+  - zanetworker
+  - sjenning
+  - slaviered
+  - xiangjingli
+approvers:
+  - csrwng
+  - zanetworker
+  - sjenning
+api-approvers: # In case of new or modified APIs or API extensions (CRDs, aggregated apiservers, webhooks, finalizers). If there is no API change, use "None"
+  - None
+creation-date: yyyy-mm-dd
+last-updated: yyyy-mm-dd
+tracking-link: # link to the tracking ticket (for example: Jira Feature or Epic ticket) that corresponds to this enhancement
+  - [OCPPLAN-5771](https://issues.redhat.com/browse/OCPPLAN-5771)
+see-also:
+replaces:
+superseded-by:
+---
+
+# Pluggable Konnectivity
+
+## Summary
+
+For worker-less deployments, hypershift deploys a konnectivity-server and -agent to facilitate seamless communication *from* the hosted control plane *to* the services running on its behalf in the hosting cluster.
+At the time of this writing, the list of services thus supported is static and hardcoded.
+
+This enhancement proposes a mechanism to facilitate adding any service to konnectivity.
+
+## Motivation
+
+One use case for hypershift -- particularly with zero-worker hosted control planes -- is to facilitate scaling of workloads, mitigating etcd limits by "offloading" kubernetes objects to etcds in one or more hosted control planes, while running the relevant workloads (e.g. controllers) on the hosting cluster.
+In such a scenario, communication *from* the hosting cluster *to* the HCP is a simple matter of creating a client with a kubeconfig pointing to the HCP.
+Communication in the other direction is almost never an issue, because any workload running "in" the HCP is actually running on the hosted cluster anyway.
+
+A notable exception is admission webhooks: The configuration (Validating- or MutatingWebhookConfiguration) on the HCP is acted on by the kube-apiservice running *on behalf of* the HCP but *actually on* the hosting cluster, which must send the requests to a Service that is similarly running on the hosting cluster.
+At the time of this writing, the only (known) way to plug an admission webhook into this configuration includes standing up a separate konnectivity-agent with the ClusterIP of the Service pointing to the webhook server.
+
+It would be preferable to be able to plug such a Service into the existing konnectivity-agent already deployed by hypershift.
+- Save duplication: In the workaround, the konnectivity-agent deployment is identical (except for IPs) to hypershift's.
+- Save cluster resources: One konnectivity-agent workload instead of two.
+- Save etcd space: Particularly in use cases where etcd slots are at a premium, we can save three objects (Deployment, ReplicaSet, Pod) by consolidating.
+- Save devtest cycles and code: Developers don't need to figure out how to run a konnectivity-agent, or include one in their project.
+
+### User Stories
+
```

Comment:
>I would like to be able to run a Service on behalf of a hosted control plane

could you be more specific about the user stories motivating this?

Reply:
To be really specific, I'm coming from hive's Scale Mode feature ([card](https://issues.redhat.com/browse/HIVE-1862), [PR](https://github.com/openshift/hive/pull/1854) -- particularly [this commit](https://github.com/openshift/hive/pull/1854/commits/04bcf358ec1c6eae7b34edfb124e69c69669e3f9) where I would love to *not* need to package my own konnectivity-agent deployment).

Reply:
The main known user story here is the case of zero worker hypershift mode with webhook support. In this case, we need to route the admission requests from the hosted control plane to the admission server pods on the hosting plane. 

The mechanism to achieve this includes adding a service and endpoint on the hosted plane that points to the admission service on the hosting plane ,and adding this clusterIP to the konnectivity agent.

The above mechanism is employed by many players such as hive, acm and assisted-installer - all of them are currently replicating the konnectivity agent.

Reply:
QQ: https://github.com/openshift/hypershift/pull/1865#issuecomment-1316800263

Reply:
[Answered](https://github.com/openshift/hypershift/pull/1865#issuecomment-1317244899).

Reply:
Adding that question & answer to the Risks and Mitigations section.